r2/types.go

```diff
@@ -0,0 +1,308 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package r2
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3metrics/rules"
+)
+
+const (
+	tagFilterListSeparator = " "
+	tagFilterSeparator     = ":"
+)
+
+// CurrentRuleSet represents the current state of a rule set. A rule with all the latest snapshots
+// of its ruleset.
+type CurrentRuleSet struct {
+	Namespace    string
+	Version      int
+	CutoverNanos int64
+	MappingRules []*rules.MappingRuleView
+	RollupRules  []*rules.RollupRuleView
+}
+
+// nolint
+func newCurrentRuleSetView(r rules.RuleSet) (*CurrentRuleSet, error) {
+	mrs, err := r.MappingRules()
+	if err != nil {
+		return nil, err
+	}
+
+	rrs, err := r.RollupRules()
+	if err != nil {
+		return nil, err
+	}
+
+	currentMappingRules := newMappingRules(mrs)
+	currentRollupRules := newRollupRules(rrs)
+
+	return &CurrentRuleSet{
+		Namespace:    string(r.Namespace()),
+		Version:      r.Version(),
+		CutoverNanos: r.CutoverNanos(),
+		MappingRules: currentMappingRules,
+		RollupRules:  currentRollupRules,
+	}, nil
+}
+
+type ruleSetJSON struct {
+	Namespace     string            `json:"id"`
+	Version       int               `json:"version"`
+	CutoverMillis int64             `json:"cutoverMillis"`
+	MappingRules  []mappingRuleJSON `json:"mappingRules"`
+	RollupRules   []rollupRuleJSON  `json:"rollupRules"`
+}
+
+func (c CurrentRuleSet) ruleSetJSON() ruleSetJSON {
+	mrJSON := make([]mappingRuleJSON, len(c.MappingRules))
+	for i, m := range c.MappingRules {
+		mrJSON[i] = newMappingRuleJSON(m)
+	}
+	rrJSON := make([]rollupRuleJSON, len(c.RollupRules))
+	for i, r := range c.RollupRules {
+		rrJSON[i] = newRollupRuleJSON(r)
+	}
+	return ruleSetJSON{
+		Namespace:     c.Namespace,
+		Version:       c.Version,
+		CutoverMillis: nanoToMilli(c.CutoverNanos),
+		MappingRules:  mrJSON,
+		RollupRules:   rrJSON,
+	}
+}
+
+// nolint
+func newMappingRules(mrs rules.MappingRules) []*rules.MappingRuleView {
+	result := make([]*rules.MappingRuleView, len(mrs))
+	idx := 0
+	for _, m := range mrs {
+		if len(m) == 0 {
+			continue
+		}
+		result[idx] = m[len(m)-1]
+		idx++
+	}
+	return result
```

Comment:
should return result[:idx]?  is there a test for this?

Is it simpler if you just use append in this case?